SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/StorageAdapter.cs

```diff
@@ -1,158 +1,203 @@
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to create table, result={result}\n\t{errorMessage}");
+            }
         }
 
-        public async Task<int> DeleteAsync<T>(Expression<Func<T, bool>> pred) where T : new()
+        public int Count(string tableName, string columnName, object value)
         {
-            try
+            var result = ExecuteSelectionSqlQuery($"SELECT COUNT(*) FROM {tableName} WHERE {columnName} = ?;", new[] { value });
+            var count = (long)(result.FirstOrDefault()?.FirstOrDefault() ?? 0L);
+            return (int)count;
+        }
+
+        public IList<object[]> Select(string tableName, string columnName, object value, string excludeColumnName, object[] excludeValues, int? limit = null)
+        {
+            var whereClause = $"{columnName} = ?";
+            var args = new List<object> { value };
+            if (excludeValues?.Length > 0)
             {
-                var numDeleted = 0;
-                var table = _dbConnection.Table<T>();
-                var entries = await table.Where(pred).ToListAsync().ConfigureAwait(false);
-                foreach (var entry in entries)
-                {
-                    numDeleted += await _dbConnection.DeleteAsync(entry).ConfigureAwait(false);
-                }
-                return numDeleted;
+                whereClause += $" AND {excludeColumnName} NOT IN ({BuildBindingMask(excludeValues.Length)})";
+                args.AddRange(excludeValues);
             }
-            catch (SQLiteException e)
+            var limitClause = limit != null ? $" LIMIT {limit}" : string.Empty;
+            var query = $"SELECT * FROM {tableName} WHERE {whereClause}{limitClause};";
+            return ExecuteSelectionSqlQuery(query, args);
+        }
+
+        public void Insert(string tableName, string[] columnNames, ICollection<object[]> values)
+        {
+            var columnsClause = string.Join(",", columnNames);
+            var valueClause = $"({BuildBindingMask(values.First().Length)})";
+            var valuesClause = string.Join(",", Enumerable.Repeat(valueClause, values.Count));
+            var valuesArray = values.SelectMany(i => i).ToArray();
+            var result = ExecuteNonSelectionSqlQuery($"INSERT INTO {tableName}({columnsClause}) VALUES {valuesClause};", valuesArray);
+            if (result != raw.SQLITE_OK)
             {
-                throw ToStorageException(e);
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to prepare insert SQL query, result={result}\n\t{errorMessage}");
             }
         }
 
-        public Task InitializeStorageAsync()
+        public void Delete(string tableName, string columnName, params object[] values)
```

Comment:
Do we actually pass objects that are not string?

Reply:
Yes. Integers to filter by ids